### ISSUE TITLE
Feature request: Option to sync word-known-level between parent and children words

### ISSUE THREAD

**Issue by Ziphilt** [jzohrab/lute-v3#66]
**Is your feature request related to a problem? Please describe.**

I'm currently studying Russian with Lute, and Russian has conjugations and declensions that lead to many, many word variations. I use Lute's word-parent feature to organize them, but it feels tedious to have to separately track how well I know each word variant, when I consider them all the same word in my mental model of the language.

**Describe the solution you'd like**

I'd like to be able to configure Lute to automatically sync the word-known-level to a parent word and all that parent's other children.

Suppose for example I'm learning English, and consider the parent word 'walk' and children 'walking', 'walked', 'walks'.
When I change how well I know 'walked', I want that change to be reflected in 'walk', 'walking', and 'walks' so they all stay in sync.

But consider the parent word 'be', with children 'is', 'am', 'are'.
Maybe I'm still getting used to the irregular forms of 'be', and I'd prefer their word-known-levels be kept separate.

So I think it would be nice to be able to opt-out of this synchronization on a per-parent basis.

Note that I don't use the feature where a word can have multiple parents. I can see how that might complicate this. I would consider it acceptable to sync word-known-level to all connected parent and sibling words as a big group, for my use cases.

**Describe alternatives you've considered**

The obvious alternative is to continue handling this manually.

As for design alternatives when trying to solve this problem, I think it's worth considering whether this should be opt-in or opt-out, and at what levels this should be configurable (global per install, or per language, or per parent word).

Maybe some users use the word-parent feature for groups of words that don't work like this, and they would prefer Lute to not try to be clever about syncing the word-known-levels.

**Additional context**

I'm really enjoying using Lute! Thanks for releasing this software open-source!

**Comment by jzohrab** [comment]
Thanks for the well-written issue, @Ziphilt . :-)

Another user asked for something similar.  Here's my initial though: when you enter a parent, it automatically sets the status for the child by default.  e.g. if you're entering term "walking" and set parent "walk" which has status 3, "walking" is auto-set to 3.

Perhaps then, when changing the status of a parent, if it sees that it has children _with the same status_, it would show a pop-up asking if you want to auto-set them to that same status.  Then I wouldn't need the extra checkbox (more screen and db clutter?).  If the children have different status (e.g. subjunctive forms in Spanish), they'd be left as-is.

Ah right, would then have to handle children having multiple parents -- hm, might need more thought.

Do you think that would work?

**Comment by Ziphilt** [comment]
Thanks for such a quick response!

Do I understand correctly, that in your proposed plan, I would need to specifically go to the parent term to change the status for all other children? I would then have to navigate to the parent for almost every word I want to change the status of (because most instances of words are some variant, not the parent form).

The popup idea sounds like a frustrating amount of clicking for what would be the default behaviour in my workflow. There are only a handful of truly irregular words I've found in Russian so far, but hundreds that look like this:

    рука
        78 руку
        33 рукой
        31 руке
        28 руки
        12 руках
        12 рука
        11 рук
         9 руками
         1 рукам

This is the output of a tool I'm making to help me find common words in my books, grouped by their lemma form (looking at the source for LuteLemma gave me a good start for using spaCy!).

The Russian word 'рука' (meaning 'hand') appears in 9 forms in this book, and the most common by far isn't even the parent form. I'm seeing quite a few other common words where the parent is also not the most common form.

All 9 of these forms are easily recognizable as 'рук*' even if I'm not familiar enough with all the different cases to produce them myself in output. So for reading purposes, I want all 9 of these treated as the same word.

If you're concerned about UI bloat for users who don't care about this feature, I don't mind having to turn something on in the configuration to have the more busy UI enabled. Though I understand that it's a permanent maintenance burden on the backend.

In summary: the proposed change isn't as convenient as I hoped, but I would still use it over manually updating all 9 variants as I slowly find them all while reading.

**Comment by jzohrab** [comment]
Ah yeah maybe my approach wasn’t good for you. To clarify:

when you create a term, you can specify the parent. During the specify, I was thinking that the new term would inherit the parent.  That would be for creation only, I thought that was good enough.

However, after creation, yes it would be independent. So if you came across that same term and changed its status, the parent would not change. Was that your concern?

re an option to update the parent with the child, or vice versa, there are some challenges I can think of, so let’s try to come up with solutions one by one.

Here’s case 1: in Spanish, verbs have different conjugations. For the most part, it’s clear, but I keep making an error in one particular situation. So I’d want all children except a few to follow the parent.  Perhaps the parent would have a “sync children” indicator, and then the naughty children :-) would have a “I am naughty” indicator.  The default for new children would be for them to be obedient. The default for new terms that might be parents in the future would be the value the user chooses in settings: “sync parent and children statuses” true or false.  If a child has multiple parents, then they’re automatically naughty.  Haha this is such a stupid metaphor but it’s entertaining.

Beyond coding complexity, I’m trying to envision the user experience. Easy to get overwhelmed :-) but the above might be ok.

Lets try to work out case 1 above before getting into other cases or scenarios. Eg a case of “force align parents and children” would be a new feature.

And maybe all of the above is overkill!  Maybe we just have an “is naughty” :-) flag on any child term, and if it’s set then their status can change. The default would be to follow the parent, period.  Less customized, but also less code. Less code = better.

Does that work?

**Comment by jzohrab** [comment]
Thinking a bit more — though not too deeply — I feel a “follow parent” checkmark or similar for the child term would be sufficient, at least for case 1.  Default is to follow.

**Comment by Ziphilt** [comment]
> Ah yeah maybe my approach wasn’t good for you. To clarify:
> 
> when you create a term, you can specify the parent. During the specify, I was thinking that the new term would inherit the parent. That would be for creation only, I thought that was good enough.
> 
> However, after creation, yes it would be independent. So if you came across that same term and changed its status, the parent would not change. Was that your concern?

Yes, exactly. I want to be able to set up the parent-child groupings in advance, using a tool like LuteLemma or manually by going through my "most common words in this book" list before I actually start reading.

When I first add the words, I'm deciding to pay attention to them, but I may be somewhere between "I might not recognize this word again" and "I know I've seen this word before but I have no idea what it means".

That is to say, I expect and plan to change the word-known-level many times, because I add words when they are quite new to me.

> re an option to update the parent with the child, or vice versa, there are some challenges I can think of, so let’s try to come up with solutions one by one.
> 
> Here’s case 1: in Spanish, verbs have different conjugations. For the most part, it’s clear, but I keep making an error in one particular situation. So I’d want all children except a few to follow the parent. Perhaps the parent would have a “sync children” indicator, and then the naughty children :-) would have a “I am naughty” indicator. The default for new children would be for them to be obedient. The default for new terms that might be parents in the future would be the value the user chooses in settings: “sync parent and children statuses” true or false. If a child has multiple parents, then they’re automatically naughty. Haha this is such a stupid metaphor but it’s entertaining.

I like this metaphor! And this workflow sounds like an improvement over my initial idea of being able to opt out, but only for a parent and _all_ its children.

This would work very well for the situation where certain conjugation/declension patterns are technically "regular" according to linguists, but I'm beginner enough in the language that I'm still getting used to the patterns.
This lets the user specifically mark which forms they are still having trouble with, while letting the others be conveniently grouped.

> Beyond coding complexity, I’m trying to envision the user experience. Easy to get overwhelmed :-) but the above might be ok.
> 
> Lets try to work out case 1 above before getting into other cases or scenarios. Eg a case of “force align parents and children” would be a new feature.
> 
> And maybe all of the above is overkill! Maybe we just have an “is naughty” :-) flag on any child term, and if it’s set then their status can change. The default would be to follow the parent, period. Less customized, but also less code. Less code = better.
> 
> Does that work?

If I understood correctly, the difference between these two hypothetical solutions is just whether parent words, or "possible parent words", have a flag on them. That lets a user configure whether to opt-in or opt-out by default.

In the second version where the parent words have no such flag, the behavior is simply that all parents have the sync behavior.
I believe this means that, if a user finds a word whose variants are irregular and they want to track them all separately, they cannot mark the parent as such, and will need to mark the irregular children as they appear.
I could see this becoming a pain in a language with _lots_ of irregular words that need to be tracked separately.
But on the other hand, learners will probably be spending lots of quality time staring at irregular words to learn them in the first place, so having to uncheck boxes for them once per word might be acceptable.

I'm happy with the simpler second option if you are.

> Thinking a bit more — though not too deeply — I feel a “follow parent” checkmark or similar for the child term would be sufficient, at least for case 1. Default is to follow. If multiple parents, can't follow. @Ziphilt - have a think through this, and see if that would suffice for you.

And to confirm: if a child and its siblings have the "follow parent" checkmark checked as per default, will it be possible to quickly edit the status of all sibling children and the parent by only editing one of the children?
That would be ideal and convenient, for the base case that I had in mind.

**Comment by jzohrab** [comment]
> And to confirm: if a child and its siblings have the "follow parent" checkmark checked as per default, will it be possible to quickly edit the status of all sibling children and the parent by only editing one of the children?

originally I had thought of disabling the children status, but I prefer your idea. Editing the child status would edit the parent and all siblings that are also following the parent.  I don’t think any warning (“hey you’re also updating parent and siblings”) is needed, because people can just edit the status again.

Ok this feels reasonable.  Need to size it and figure out when to do it. Can’t commit to any time but I agree this is a good change so I’d like to do it sooner than some of the other things in the pipeline.  Cheers!

**Comment by Ziphilt** [comment]
> > And to confirm: if a child and its siblings have the "follow parent" checkmark checked as per default, will it be possible to quickly edit the status of all sibling children and the parent by only editing one of the children?
> 
> originally I had thought of disabling the children status, but I prefer your idea. Editing the child status would edit the parent and all siblings that are also following the parent. I don’t think any warning (“hey you’re also updating parent and siblings”) is needed, because people can just edit the status again.

Awesome, I'm excited for how much easier it will be to deal with these word variants!

> Ok this feels reasonable. Need to size it and figure out when to do it. Can’t commit to any time but I agree this is a good change so I’d like to do it sooner than some of the other things in the pipeline. Cheers!

Excellent, cheers!

**Comment by jzohrab** [comment]
Thinking more about this feature, to get an idea of what's needed for release.  In addition to the idea outlined above, there will need to be some kind of "clean-up" feature on initial launch for users (with lots of terms and parents) to clean up their data.

I think a simple data export would suffice, exporting all parents and children that aren't following their parent statuses.  Users could then import that CSV into something and indicate which children should follow their parents.  Then they'd import that file, and things would be linked up correctly.

**Comment by jzohrab** [comment]
Have been working on a related piece for this, it’s in progress.

**Comment by jzohrab** [comment]
Further code specs, just storing here for safekeeping.  This should go into some technical doc or something.  The below outline will inform the automated tests (unit and acceptance), and some user documentation.

## Existing data cleanup

Some users (like me) already have a bunch of data, so Lute will provide some kind of "export to CSV" thing that users can then load into excel or google sheets and bulk-clean-up their data.  The cleaned up data will then be re-importable into Lute and the data fixed with the new settings.

Export should include:
- Single parent terms only
- "Unlinked" terms only  (no terms will be linked at first)

For import, a bunch of tests are needed:
- Changes single parent only
- Headings must be correct
- Lang match
- Inconsistent parent statuses throws
- If linked child gets parent status only
- Empty parent, parent status, child, child status, lang throws 
- Y or y in "follow_parent" field means follow parent
- Can change parent status
- Can change child status if unlinked
- Throw if parent - child link in db is different than what’s in file  (e.g. between time of export and import, user may have changed the parentage in the db)

## Model/db level changes

Model changes, unit tests
- No parent or multi parent - unlink all
- Single parent- save current value to parent and siblings
- Save parent updates linked children, not unlinked
- deleting parent unlinks all children

## Actual UI work

On load:
- Single parent - enable "follow_parent" checkbox
- Multi or none: disable, clear "follow_parent" checkbox"

If set parent tag changes:
    ⁃    no parent is disable checkbox, uncheck
    ⁃    Multi is disable, uncheck
    ⁃    Single is enable, check

**Comment by jzohrab** [comment]
Has been merged into the `develop` branch, have a few minor tweaks and some docs to write for the manual, and then will do a release.

**Comment by jzohrab** [comment]
Launched in v 3.1.0, making the release now.  Docs are at https://jzohrab.github.io/lute-manual/usage/terms/link-parent-child-statuses.html.

**Comment by jzohrab** [comment]
... and launched.  I'm expecting a fair amount of feedback/pushback on this one. :-)  But it's a step forward.  @Ziphilt thanks for the notes and discussion.  Cheers!
